lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1466,61 +1463,82 @@ computeAutomaticEnumValueKind(TypeChecker &TC, EnumDecl *ED) {
   }
 }
 
-static void checkEnumRawValues(TypeChecker &TC, EnumDecl *ED) {
+llvm::Expected<bool>
+EnumRawValuesRequest::evaluate(Evaluator &eval, EnumDecl *ED,
+                               TypeResolutionStage stage) const {
   Type rawTy = ED->getRawType();
   if (!rawTy) {
-    return;
+    return true;
   }
-
+  
   if (ED->getGenericEnvironmentOfContext() != nullptr)
     rawTy = ED->mapTypeIntoContext(rawTy);
   if (rawTy->hasError())
-    return;
-
-  // If we don't have a value kind, the decl checker will provide a diagnostic.
-  auto valueKind = computeAutomaticEnumValueKind(TC, ED);
-  if (!valueKind.hasValue())
-    return;
-  
+    return true;
+    
   // Check the raw values of the cases.
   LiteralExpr *prevValue = nullptr;
   EnumElementDecl *lastExplicitValueElt = nullptr;
 
   // Keep a map we can use to check for duplicate case values.
   llvm::SmallDenseMap<RawValueKey, RawValueSource, 8> uniqueRawValues;
 
+  auto uncheckedRawValueOf = [](EnumElementDecl *EED) -> LiteralExpr * {
+    return EED->RawValueExpr;
+  };
+  
+  auto *TC = static_cast<TypeChecker *>(ED->getASTContext().getLazyResolver());
+  Optional<AutomaticEnumValueKind> valueKind;
   for (auto elt : ED->getAllElements()) {
-    // Skip if the raw value expr has already been checked.
-    if (elt->hasRawValueExpr() && elt->getRawValueExpr()->getType()) {
-      prevValue = elt->getRawValueExpr();
-      continue;
-    }
-
-    // Make sure the element is checked out before we poke at it.
-    // FIXME: Make isInvalid work with interface types
-    (void)elt->getInterfaceType();
+    // If the element has been diagnosed up to now, skip it.
     if (elt->isInvalid())
       continue;
     
-    if (elt->hasRawValueExpr()) {
-      lastExplicitValueElt = elt;
-    } else {
+    if (uncheckedRawValueOf(elt)) {
+      if (!uncheckedRawValueOf(elt)->isImplicit())
+        lastExplicitValueElt = elt;
+    } else if (!ED->LazySemanticInfo.hasFixedRawValues()) {
+      assert(TC && "Must have a lazy resolver set");
+      // Try to pull out the automatic enum value kind.  If that fails, bail.
+      if (!valueKind) {
+        valueKind = computeAutomaticEnumValueKind(*TC, ED);
+        if (!valueKind) {
+          elt->setInvalid();
+          return true;
+        }
+      }
+      
       // If the enum element has no explicit raw value, try to
       // autoincrement from the previous value, or start from zero if this
       // is the first element.
-      auto nextValue = getAutomaticRawValueExpr(TC, *valueKind, elt, prevValue);
+      auto nextValue = getAutomaticRawValueExpr(*TC, *valueKind, elt, prevValue);
```

Comment:
You can refactor getAutomaticRawValueExpr() too:

- Get ASTContext from the Decl instead of from TC.Context
- Use ctx.Diags.diagnose() instead of TC.diagnose()